[quant][core][improvement] Added warnings to quantized dynamic conv and linear ops when reduce_range=true

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79273

Summary:
Previously, the backend code "silently" ignores reduce_range=true when
using the qnnpack backend (which does not require a reduction in range).
We evaluated either 1) respecting the reduction in range to conform with
other backends (e.g., fbgemm) even when qnnpack does support the full
range and outputting a warning to let the user know that reduce_range
shoudl be set to false for qnnpack backend 2) throwing a warning and letting the user know that the
reduce_range=true setting is being ignored.

Option 1 would halve the range which could have some negative
implications to accuracy and lead to bc-breaking changes. Option 2 is also not ideal because it ignores any user settings
for reduce_range=true when using the qnnpack backend with dynamic and
linear quantized ops. We decided to go with option 2 as it is not
bc-breaking.

Fixes #68278

Test plan:

Differential Revision: [D37068897](https://our.internmc.facebook.com/intern/diff/D37068897)

dzdang committed:
[quant][core][improvement] Added warnings for when reduce_range is set to true when using qnnpack backend

Summary:
Previously, the backend code "silently" ignores reduce_range=true when
using the qnnpack backend (which does not require a reduction in range).
We evaluated either 1) respecting the reduction in range to conform with
other backends (e.g., fbgemm) even when qnnpack does support the full
range 2) throwing a warning and letting the user know that the
reduce_range=true setting is being ignored.

Option 1 would halve the range which could have some negative
implications to accuracy, so we have opted for option 2, which has the
negative consequence of overriding the user's input. If there is a
strong objection to this, we could explore option 1 again, or allow the
user to choose either option.

Fixes #68278

Test plan:

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37068897).

vkuzo reviewed (changes requested):
it's unexpected to ignore a user setting, can we honor the setting and change the warning to let the user know there is a better way?

dzdang committed:
Update on "[quant][core][improvement] Added warnings for when reduce_range is set to true when using qnnpack backend"


Summary:
Previously, the backend code "silently" ignores reduce_range=true when
using the qnnpack backend (which does not require a reduction in range).
We evaluated either 1) respecting the reduction in range to conform with
other backends (e.g., fbgemm) even when qnnpack does support the full
range 2) throwing a warning and letting the user know that the
reduce_range=true setting is being ignored.

Option 1 would halve the range which could have some negative
implications to accuracy, so we have opted for option 2, which has the
negative consequence of overriding the user's input. If there is a
strong objection to this, we could explore option 1 again, or allow the
user to choose either option.

Fixes #68278


Differential Revision: [D37068897](https://our.internmc.facebook.com/intern/diff/D37068897)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37068897).

dzdang committed:
Update on "[quant][core][improvement] Made quantized conv and linear ops use reduce_range when using qnnpack backend"


Summary:
Previously, the backend code "silently" ignores reduce_range=true when
using the qnnpack backend (which does not require a reduction in range).
We evaluated either 1) respecting the reduction in range to conform with
other backends (e.g., fbgemm) even when qnnpack does support the full
range and outputting a warning to let the user know that reduce_range
shoudl be set to false for qnnpack backend 2) throwing a warning and letting the user know that the
reduce_range=true setting is being ignored.

Option 1 would halve the range which could have some negative
implications to accuracy; however, we believe it is more important to
respect what the user is requesting and there have chosen to implement this option.

Fixes #68278

Test plan:

Differential Revision: [D37068897](https://our.internmc.facebook.com/intern/diff/D37068897)

[ghstack-poisoned]

jerryzh168 commented:
@vkuzo we definitely should make this better but honoring the option would be bc-breaking right now, maybe we should not make bc-breaking changes until we finalize the plan for quantized engines? wdyt?

dzdang commented:
@vkuzo @jerryzh168 I tried honoring `reduce_range=True` for qnnpack, and it led to a lot of test failures (see current failures in the CI, a lot of it are from jit). I dug a bit deeper into some of the test failures and found that a lot of these tests (and our workflow) assume qnnpack never reduces the range, so this would be a bit tricky to fix all of these inherent assumptions

vkuzo commented:
makes sense, then can we make the warning say something like "currently, qnnpack incorrectly ignores reduce_range, this may change in a future release"?

dzdang committed:
Update on "[quant][core][improvement] Added warnings to quantized dynamic conv and linear ops when reduce_range=true"


Summary:
Previously, the backend code "silently" ignores reduce_range=true when
using the qnnpack backend (which does not require a reduction in range).
We evaluated either 1) respecting the reduction in range to conform with
other backends (e.g., fbgemm) even when qnnpack does support the full
range and outputting a warning to let the user know that reduce_range
shoudl be set to false for qnnpack backend 2) throwing a warning and letting the user know that the
reduce_range=true setting is being ignored.

Option 1 would halve the range which could have some negative
implications to accuracy and lead to bc-breaking changes. Option 2 is also not ideal because it ignores any user settings
for reduce_range=true when using the qnnpack backend with dynamic and
linear quantized ops. We decided to go with option 2 as it is not
bc-breaking.

Fixes #68278

Test plan:

Differential Revision: [D37068897](https://our.internmc.facebook.com/intern/diff/D37068897)

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp:
```diff
@@ -72,6 +72,11 @@ template <int kSpatialDim>
     bool /*reduce_range*/) {
+  if (reduce_range) {
+    TORCH_WARN("Currently, qnnpack incorrectly ignores reduce_range when it is set to true; this may change in a future release. "
+    "Reducing the range for qnnpack would currently lead to bc-breaking behavior.");
```
nit: maybe we can remove this line?

dzdang committed:
Update on "[quant][core][improvement] Added warnings to quantized dynamic conv and linear ops when reduce_range=true"


Summary:
Previously, the backend code "silently" ignores reduce_range=true when
using the qnnpack backend (which does not require a reduction in range).
We evaluated either 1) respecting the reduction in range to conform with
other backends (e.g., fbgemm) even when qnnpack does support the full
range and outputting a warning to let the user know that reduce_range
shoudl be set to false for qnnpack backend 2) throwing a warning and letting the user know that the
reduce_range=true setting is being ignored.

Option 1 would halve the range which could have some negative
implications to accuracy and lead to bc-breaking changes. Option 2 is also not ideal because it ignores any user settings
for reduce_range=true when using the qnnpack backend with dynamic and
linear quantized ops. We decided to go with option 2 as it is not
bc-breaking.

Fixes #68278

Test plan:

Differential Revision: [D37068897](https://our.internmc.facebook.com/intern/diff/D37068897)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37068897).

dzdang committed:
Update on "[quant][core][improvement] Added warnings to quantized dynamic conv and linear ops when reduce_range=true"


Summary:
Previously, the backend code "silently" ignores reduce_range=true when
using the qnnpack backend (which does not require a reduction in range).
We evaluated either 1) respecting the reduction in range to conform with
other backends (e.g., fbgemm) even when qnnpack does support the full
range and outputting a warning to let the user know that reduce_range
shoudl be set to false for qnnpack backend 2) throwing a warning and letting the user know that the
reduce_range=true setting is being ignored.

Option 1 would halve the range which could have some negative
implications to accuracy and lead to bc-breaking changes. Option 2 is also not ideal because it ignores any user settings
for reduce_range=true when using the qnnpack backend with dynamic and
linear quantized ops. We decided to go with option 2 as it is not
bc-breaking.

Fixes #68278

Test plan:

Differential Revision: [D37068897](https://our.internmc.facebook.com/intern/diff/D37068897)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37068897).

dzdang commented:
@pytorchbot merge -g

dzdang commented:
not sure why this didn't get auto merged when the internal diff landed. going to try to merge it manually
